src/bootstrap.jl

```diff
@@ -1,95 +1,150 @@
 end
 
-function BootstrapBernoulli{S <: ContinuousUnivariateStreamStat}(
+# Double or nothing bootstrap
+function BernoulliBootstrap{S <: ContinuousUnivariateStreamStat}(
     stat::S,
     R::Int = 1_000,
     α::Real = 0.05,
 )
-    replicates = S[copy(stat) for i in 1:R]
-    replicate_states = Array(Float64, R)
-    return BootstrapBernoulli(replicates, replicate_states, α, 0)
+    stat_replicates = S[copy(stat) for i in 1:R]
+    replicates = Array(Float64, R)
+    return BernoulliBootstrap(stat_replicates, replicates, α, 0, true)
 end
 
-type BootstrapPoisson{S <: ContinuousUnivariateStreamStat} <: Bootstrap
-    replicates::Vector{S}
-    replicate_states::Vector{Float64}
-    α::Float64
-    n::Int
-    # TODO: Force vectors to both have size R
-end
-
-function BootstrapPoisson{S <: ContinuousUnivariateStreamStat}(
+function PoissonBootstrap{S <: ContinuousUnivariateStreamStat}(
     stat::S,
     R::Int = 1_000,
     α::Float64 = 0.05,
 )
-    replicates = S[copy(stat) for i in 1:R]
-    replicate_states = Array(Float64, R)
-    return BootstrapPoisson(replicates, replicate_states, α, 0)
+    stat_replicates = S[copy(stat) for i in 1:R]
+    replicates = Array(Float64, R)
+    return PoissonBootstrap(stat_replicates, replicates, α, 0, true)
 end
 
-function update!(stat::BootstrapBernoulli, args::Any...)
+function update!(stat::BernoulliBootstrap, args::Any...)
     stat.n += 1
 
-    for replicate in stat.replicates
+    for stat_replicate in stat.stat_replicates
         if rand() > 0.5
-            update!(replicate, args...)
-            update!(replicate, args...)
+            update!(stat_replicate, args...)
+            update!(stat_replicate, args...)
         end
     end
-
+    stat.dirty = true
     return
 end
 
-function update!(stat::BootstrapPoisson, args::Any...)
+function update!(stat::PoissonBootstrap, args::Any...)
     stat.n += 1
 
-    for replicate in stat.replicates
+    for stat_replicate in stat.stat_replicates
         repetitions = rand(Distributions.Poisson(1))
         for repetition in 1:repetitions
-            update!(replicate, args...)
+            update!(stat_replicate, args...)
         end
     end
+    stat.dirty = true
 
     return
 end
 
-# TODO: Make this work with non-univariate statistics by taking marginal
-#       quantiles
-function state(stat::Bootstrap)
-    R = length(stat.replicates)
+function update!(stat::Bootstrap, args::Any...)
+    error("Cannot update a FrozenBootstrap object")
+    return
+end
 
-    for (i, replicate) in enumerate(stat.replicates)
-        stat.replicate_states[i] = state(replicate)
+# update replicates' states if necessary and return their values
+function replicates!(stat::Bootstrap)
+    if stat.dirty
+        for (i, stat_replicate) in enumerate(stat.stat_replicates)
+            stat.replicates[i] = state(stat_replicate)
+        end
+        stat.dirty = false
     end
+    return stat.replicates
+end
+
+function replicates!(stat::FrozenBootstrap)
```

Comment:
In Julia, functions that end in `!` are meant to do mutation. Since this is a field getter, it's not clear to me that it should have an exclamation mark.